### PR TITLE
Coll updateMany & patch implementations

### PR DESCRIPTION
**Pull request by ApexTheory**
PR for https://github.com/ergoplatform/sigma-rust/issues/357 and https://github.com/ergoplatform/sigma-rust/issues/359

### CODE REVIEW

File: ergotree-ir/README.md [hunk 1]
```diff
@@ -199,7 +195,9 @@ Descriptions for the operations can be found in [ErgoTree Specification](https:/
 - indexOf
 - zip [#329](https://github.com/ergoplatform/sigma-rust/issues/329)
 - indices [#314](https://github.com/ergoplatform/sigma-rust/issues/314)
+- patch [#357](https://github.com/ergoplatform/sigma-rust/issues/357)
 - updated [#358](https://github.com/ergoplatform/sigma-rust/issues/358)
+- updateMany [#359](https://github.com/ergoplatform/sigma-rust/issues/359)
```
Comment (greenhat) [inline ergotree-ir/README.md:200]: Thanks for updating this!

File: ergotree-interpreter/src/eval/scoll.rs [hunk 2]
```diff
@@ -151,6 +151,60 @@ pub(crate) static INDICES_EVAL_FN: EvalFn = |_env, _ctx, obj, _args| {
...
+        .cloned()
+        .ok_or_else(|| EvalError::NotFound("patch: missing first arg (from)".to_string()))?;
+    let patch_val = args
+        .get(1)
+        .cloned()
+        .ok_or_else(|| EvalError::NotFound("patch: missing second arg (patch)".to_string()))?;
+    let replaced_val = args
+        .get(2)
+        .cloned()
+        .ok_or_else(|| EvalError::NotFound("patch: missing second arg (replaced)".to_string()))?;
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/scoll.rs:173]:
```suggestion
        .ok_or_else(|| EvalError::NotFound("patch: missing third arg (replaced)".to_string()))?;
```

File: ergotree-interpreter/src/eval/scoll.rs [hunk 3]
```diff
@@ -151,6 +151,60 @@ pub(crate) static INDICES_EVAL_FN: EvalFn = |_env, _ctx, obj, _args| {
...
+    i = replaced;
+    while i > 0 && iter.peek().is_some() {
+        iter.next();
+        i -= 1;
+    }
+    while iter.peek().is_some() {
+        if let Some(elem) = iter.next().cloned() {
+            res.push(elem)
+        }
+    }
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/scoll.rs:204]: How do you like to ditch those three loops above in favor of something like this:
```
let res = normalized_input_vals.iter()
    .take(from)
    .chain(patch)
    .chain(normalized_input_vals.iter().skip(from+replaced))
    .collect()
```
Please check for one-off errors though.

File: ergotree-interpreter/src/eval/scoll.rs [hunk 4]
```diff
@@ -184,6 +238,75 @@ pub(crate) static UPDATED_EVAL_FN: EvalFn = |_env, _ctx, obj, args| {
...
+        let updates_len = updates_val.len();
+
+        if indexes_len != updates_len {
+            return Err(EvalError::UnexpectedValue(format!(
+                "Collections should have same length but was: \
+            {0:?} and {1:?}. \n Indexes: {2:?} \n Updates: {3:?}",
+                indexes_len, updates_len, indexes_usize, updates_val
+            )));
+        };
+
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/scoll.rs:283]: Let's also check that `updates.tpe == inputs.tpe`.

File: ergotree-interpreter/src/eval/scoll.rs [hunk 5]
```diff
@@ -184,6 +238,75 @@ pub(crate) static UPDATED_EVAL_FN: EvalFn = |_env, _ctx, obj, args| {
...
+                    "updateMany index out of bounds, got: {0:?}",
+                    pos
+                )));
+            }
+            let update = updates_val[i].clone();
+            match res.get_mut(pos) {
+                Some(elem) => *elem = update,
+                None => {
+                    return Err(EvalError::UnexpectedValue(format!(
+                        "updateMany index not found, got: {0:?}",
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/scoll.rs:300]:
```suggestion
                        "updateMany index out of bounds, got: {0:?}",
```
I mean, I don't see any other reason for it to fail.

**Review by greenhat — CHANGES_REQUESTED** [review]
Looking great! I dig the tests! Please check the comments for minor notes.

**Comment by ApexTheory** [comment]
Updated with your suggestions.

**Review by greenhat — APPROVED** [review]
Thank you!

**Pull request merged**